File: src/main/java/org/aopbuddy/plugin1/infra/util/PortUtil.java
package org.aopbuddy.plugin1.infra.util;

import java.io.IOException;
import java.net.ServerSocket;

public class PortUtil {
    public static int findFreePort() {
        try (ServerSocket serverSocket = new ServerSocket(0)) {
            // 绑定到端口 0，系统会自动分配一个空闲端口
//            当一个服务器程序（如 ServerSocket）绑定到一个端口并运行一段时间后关闭，操作系统不会立即释放该端口。端口可能会进入 TIME_WAIT 状态（通常持续 30 秒到 2 分钟，具体取决于操作系统）
//            设置 setReuseAddress(true) 允许 ServerSocket 绑定到这个端口
            serverSocket.setReuseAddress(true); // 允许端口快速重用
            return serverSocket.getLocalPort(); // 获取分配的端口
        } catch (IOException e) {
            throw new RuntimeException("Failed to find a free port", e);
        }
    }
}

File: src/main/kotlin/org/aopbuddy/plugin/infra/util/PortUtil.kt
package org.aopbuddy.plugin.infra.util

import java.io.IOException
import java.net.ServerSocket

object PortUtil {
    fun findFreePort(): Int {
        // 从8888到8900依次遍历寻找可用端口
        for (port in 8888..8900) {
            try {
                ServerSocket(port).use { serverSocket ->
                    // 设置端口快速重用
                    serverSocket.reuseAddress = true
                    // 如果能成功创建ServerSocket，说明端口可用
                    return port
                }
            } catch (e: IOException) {
                // 端口被占用，继续尝试下一个端口
                continue
            }
        }
        // 如果8888-8900端口都被占用，抛出异常
        throw RuntimeException("Failed to find a free port in range 8888-8900")
    }
}